### PR TITLE
Clean up Cloud Storage backend

### PR DESCRIPTION
**Pull request by wgfm**
This MR cleans up the Cloud Storage backend code. All HTTP related code is now in single module (gcs_client). There should be no functional changes to the code.

Most duplication is removed. In particular:
- The Authorization header is now set in a single function, rather than in every request that we build
- Handling trailing slashes in URLs for files vs directories is now more clear, due to the `TrailingSlash` enum.
- JSON deserialization is now done in a single place using `gcs_client::deserialize`, which is generic over its return type.

All token logic now lives into a single TokenManager, which handles fetching and caching tokens.

TODO:
- [ ] Figure out how to test gcs_client, as the existing URL tests can't run in the current state. Maybe I could write a Hyper connector for recording Requests, or something similar.

### CODE REVIEW

**Comment by hannesdejager** [comment]
Nice, please rebase or latest master so the new pipelines can run. I'll check this out soonish.

**Comment by robklg** [comment]
Looks like a big improvement. Thanks a lot. I'm going to test things coming week.

File: crates/unftp-sbe-gcs/src/lib.rs [hunk 1]
```diff
@@ -58,53 +58,32 @@
 // FIXME: error mapping from GCS/hyper is minimalistic, mostly PermanentError. Do proper mapping and better reporting (temporary failures too!)
```
Comment (hannesdejager) [inline crates/unftp-sbe-gcs/src/lib.rs:58]: I think remove this FIXME too and lets just make a ticket if anyone have an actual plan.

File: crates/unftp-sbe-gcs/src/lib.rs [hunk 2]
```diff
@@ -256,77 +132,29 @@ impl<User: UserDetail> StorageBackend<User> for CloudStorage {
...
-
-        let response: Response<Body> = client.request(request).map_err(|e| Error::new(ErrorKind::PermanentFileNotAvailable, e)).await?;
-
-        let body = unpack_response(response).await?;
-
-        let response: Item = serde_json::from_reader(body.reader()).map_err(|e| Error::new(ErrorKind::PermanentFileNotAvailable, e))?;
-
-        response.to_metadata()
+    async fn metadata<P>(&self, _user: &User, path: P) -> Result<Self::Metadata, Error>
+    where
```
Comment (hannesdejager) [inline crates/unftp-sbe-gcs/src/lib.rs:136]: yeah we can more consistently use where clauses. Its more ergonomic mostly.

File: crates/unftp-sbe-gcs/src/uri.rs [hunk 3]
```diff
@@ -1,210 +0,0 @@
-use hyper::Uri;
-use libunftp::storage::{Error, ErrorKind};
-use percent_encoding::{utf8_percent_encode, NON_ALPHANUMERIC};
-use std::path::{Path, PathBuf};
-
-#[derive(Clone, Debug)]
-pub(crate) struct GcsUri {
```
Comment (hannesdejager) [inline crates/unftp-sbe-gcs/src/uri.rs:7]: Curious why you would combine the GcsUri stuff with the client if it was already separate? what is your thinking?

File: crates/unftp-sbe-gcs/src/gcs_client.rs [hunk 4]
```diff
@@ -0,0 +1,529 @@
...
+    response_body::{Item, ResponseBody},
+    workload_identity,
+};
+
+type HttpClient = Client<HttpsConnector<HttpConnector>>;
+
+#[derive(Clone, Debug)]
+pub(crate) struct GcsClient {
+    base_url: String,
+    bucket: String,
```
Comment (hannesdejager) [inline crates/unftp-sbe-gcs/src/gcs_client.rs:31]: `bucket_name` ?

File: crates/unftp-sbe-gcs/src/gcs_client.rs [hunk 5]
```diff
@@ -0,0 +1,529 @@
...
+
+    serde_json::from_reader(body.reader()).map_err(|e| Error::new(ErrorKind::PermanentFileNotAvailable, e))
+}
+
+fn make_uri(path_and_query: String) -> Result<Uri, Error> {
+    Uri::from_maybe_shared(path_and_query).map_err(|_| Error::from(ErrorKind::FileNameNotAllowedError))
+}
+
+#[derive(Clone, Debug)]
+struct TokenManager {
```
Comment (hannesdejager) [inline crates/unftp-sbe-gcs/src/gcs_client.rs:304]: I hate the word manager, it never indicates what something is actually doing. If we can't find a better name then lets at least add a description indicating its intent.

File: crates/unftp-sbe-gcs/src/lib.rs [hunk 6]
```diff
@@ -71,7 +72,7 @@ use bytes::Buf;
 use futures::{prelude::*, TryStreamExt};
 use hyper::{
     body::aggregate,
-    client::connect::{dns::GaiResolver, HttpConnector},
+    client::connect::{HttpConnector},
```
Comment (hannesdejager) [inline crates/unftp-sbe-gcs/src/lib.rs]: Remove brackets, clippy will complain.

**Comment by hannesdejager** [comment]
A good effort thank you @wgfm. It looks a lot cleaner. I'm just worried about the testing, whats your plan to ensure all is still good?

File: crates/unftp-sbe-gcs/src/gcs_client.rs (hunk 4 above)
Comment (wgfm) [reply crates/unftp-sbe-gcs/src/gcs_client.rs]: Sure and done

File: crates/unftp-sbe-gcs/src/gcs_client.rs (hunk 5 above)
Comment (wgfm) [reply crates/unftp-sbe-gcs/src/gcs_client.rs]: Yeah, me too. I took this name from another Google APIs client. I renamed it to TokenSource. Let me know whether that's better or not

File: crates/unftp-sbe-gcs/src/lib.rs (hunk 6 above)
Comment (wgfm) [reply crates/unftp-sbe-gcs/src/lib.rs]: Hmm, I run format on save which usually fixes this for me. Perhaps it didn't pick this one up.

File: crates/unftp-sbe-gcs/src/uri.rs (hunk 3 above)
Comment (wgfm) [reply crates/unftp-sbe-gcs/src/uri.rs:7]: I wanted to move all HTTP related code to one place, so I merged the request logic with GcsUri, and then renamed GcsUri to GcsClient. I never really liked the way I set up GcsUri in the first place - I think I created the wrong abstraction then.

If you do think this is a valuable abstraction to have, I can split the URI construction from the request logic. I don't think this is common practice, though.

**Comment by wgfm** [comment]
About testing: My current line of thinking is to write a Hyper test connector that records a request, and returns a predetermined response. We'd have to make GcsClient generic over the http connector, which I've had problems with in the past, but maybe I can get it to work this time.

If that doesn't work, I can try Go's httptest approach and run an in-process http server that I can talk to with the GcsClient.

File: crates/unftp-sbe-gcs/src/uri.rs (hunk 3 above)
Comment (hannesdejager) [reply crates/unftp-sbe-gcs/src/uri.rs:7]: No lets leave it as you have it now.

File: crates/unftp-sbe-gcs/src/gcs_client.rs (hunk 5 above)
Comment (hannesdejager) [reply crates/unftp-sbe-gcs/src/gcs_client.rs]: Much better already, thanks

**Comment by hannesdejager** [comment]
From a code perspective I'm happy but I'll let Rob decide when to merge as he mentioned he may want to test it a bit.

**Pull request merged**